perf_bench_test.go

```diff
@@ -118,83 +130,122 @@ func TestBenchRequestPerSecond(t *testing.T) {
-	go profileControl(vmNum, &isProfile, profiler)
+	for i := 1.; i < 9; i++ {
+		var (
+			vmID, requestID             int
+			invokExecTime, realRequests int64
+			serveMetric                 = metrics.NewMetric()
+			rps                         = 0.1 * i * float64(targetRPS)
+			timeInterval                = time.Duration(time.Second.Nanoseconds() / int64(rps))
+			ticker                      = time.NewTicker(timeInterval)
+			totalRequests               = injectDuration * rps
+			remainingRequests           = totalRequests
+			profiler                    = profile.NewProfiler(injectDuration, *profilerInterval, vmNum, *profilerLevel, *profilerNodes, "profile", *isBindSocket)
+		)
 
-	for remainingRequests > 0 {
-		if tickerT := <-ticker.C; !tickerT.IsZero() {
-			vmGroup.Add(1)
-			remainingRequests--
+		log.Debugf("RPS: %f", rps)
+
+		tStart := time.Now()
+		latencies := make(chan float64, 2)
+		go latencyMeasurement(t, 1, images[1%len(images)], &isProfile, latencies)
+
+		err := profiler.Run()
+		require.NoError(t, err, "Run profiler returned error")
+		go profileControl(vmNum, &isProfile, profiler)
+
+		for remainingRequests > 0 {
+			if tickerT := <-ticker.C; !tickerT.IsZero() {
+				vmGroup.Add(1)
+				remainingRequests--
+
+				imageName := images[vmID%len(images)]
 
-			imageName := images[vmID%len(images)]
+				go serveVM(t, &vmGroup, vmID, requestID, imageName, isSyncOffload, &isProfile, &invokExecTime, &realRequests)
+				requestID++
 
-			go serveVM(t, &vmGroup, vmID, requestID, imageName, isSyncOffload, &isProfile, &invokExecTime, &realRequests)
-			requestID++
+				vmID = (vmID + 1) % vmNum
+			}
+		}
+		ticker.Stop()
+		vmGroup.Wait()
+		log.Debugf("All VM returned in %d Milliseconds", time.Since(tStart).Milliseconds())
+		meanLatency, percentileLatency := <-latencies, <-latencies
+		log.Debug(meanLatency, percentileLatency)
+
+		// Collect results
+		serveMetric.MetricMap[averageExecutionTime] = float64(invokExecTime) / float64(realRequests)
+		serveMetric.MetricMap[realRequestsPerSecond] = float64(realRequests) / (profiler.GetCoolDownTime() - profiler.GetWarmUpTime())
+		if cores > vmNum {
+			serveMetric.MetricMap[realRequestsPerSecond] /= float64(vmNum)
+		} else {
+			serveMetric.MetricMap[realRequestsPerSecond] /= float64(cores)
+		}
+		result, err := profiler.GetResult()
+		profiledCores := profiler.GetCores()
+		log.Debugf("%d cores are recorded: %v", len(profiledCores), profiledCores)
+		require.NoError(t, err, "Stopping profiler returned error: %v", err)
+		for eventName, value := range result {
+			log.Debugf("%s: %f\n", eventName, value)
+			serveMetric.MetricMap[eventName] = value
+		}
+		log.Debugf("average-execution-time: %f\n", serveMetric.MetricMap[averageExecutionTime])
+		log.Debugf("real-requests-per-second: %f\n", serveMetric.MetricMap[realRequestsPerSecond])
+		profiler.PrintBottlenecks()
 
-			vmID = (vmID + 1) % vmNum
+		if percentileLatency > float64(threshold) {
+			if pmuMetric.MetricMap == nil {
```

Comment:
I don't get 240-243, can you add comments?

Reply:
It breaks the loop if tail latency is violated.